### DISCUSSION TITLE
关于[中文混合拼音]的长文本OCR方案请教（eg: 灿烂的笑róng）

### DISCUSSION THREAD

**Discussion by robotJie** [PaddlePaddle/PaddleOCR#13132] [Q&A]
### Discussed in https://github.com/PaddlePaddle/PaddleOCR/discussions/13106

<div type='discussions-op-text'>

<sup>Originally posted by **robotJie** June 17, 2024</sup>
如题，要识别的样例图片如下：
![微信图片_20231205132817](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/c3548e19-7d14-4efe-9ea1-902e73da0af3)

### ------ 已经做的尝试 -------

**总体思路是：找一个已有的可识别中英文的model，再用拼音数据去做finetune。**

然后参考的这个项目[基于PaddleOCR的小学生手写汉语拼音识别](https://aistudio.baidu.com/projectdetail/5517190?channelType=0&channel=0)，用的预训练模型和配置如图
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/7ecca546-eff9-4135-86b7-e7c909642677)
配置文件除了必要的数据路径，其它基本都没动。数据用的参考项目的。

训练完之后，首先用训练集的数据测试，都能正确识别。但是用新的数据，无法正确识别。比如
infer_img为：
![yin - 副本](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/91f2747b-c3a8-4fc5-810b-3ff6cd231cd1)
，输出结果为：
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/7e821779-5964-4536-bb51-a5238e198d80)

### ------ 我的问题 ------

1. 出现上面这种情况，一种可能原因是不是过拟合？我的epoch用的默认800，但是数据集只有500（所以800是否过大了？）
2. 这个思路（_找一个已有的可识别中英文的model，再用拼音数据去做finetune_）是否可行？是否有更推荐的预训练model，比如v3？或者更好的实现方案呢？
3. 我用官方的预训练模型，去测试一个很正常的图片，结果也是错的，这是为什么？
  `python tools/infer_rec.py -c ch_PP-OCRv2_rec.yml -o Global.pretrained_model=pretrain_models/ch_PP-OCRv2_rec_slim_quant_train/best_accuracy Global.infer_img=./train_data/pinyin_image_data/self_data/beizhu.png`

    ![beizhu](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/cefb4f93-d65b-4093-9dfe-1eb6b1ff580d) beizhu.png
  
    输出结果（每次识别结果不同）：
  ![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/8bdf68d8-b95c-49f8-88f6-fa1289f6948c)

4.我发现对于大段的文字，比如像样例图片这种，识别结果只会输出零星的几个字，是需要对文字检测作什么调整么？

**---------------  任何回复，不甚感激！ ----------------**
</div>

**Comment by UserWangZz** [comment]
问题1. 500张图片的数据集对于识别任务来说有点少。
问题2. 使用预训练模型对拼音数据进行finetune是可以的，针对问题提到的情况，建议考虑增加数据量，即将发布的PPChatOCR v3将带来全新的融合数据训练机制，可以一定程度上保证模型在通用数据集上的精度，同时在垂类数据集上保证可行的精度。
问题3. 根据所提供的信息，感觉可能是识别的字典没有匹配上的感觉。
问题4. 样例图片中，可能作文的格子影响了检测模型的检测，导致识别效果不佳，也有可能检测结果可以，但是因为格子的竖线影响了识别模型的精度，导致最后被过滤掉。

**Reply by running-az** [reply]
请问哪个模型有对拼音的识别？

**Reply by UserWangZz** [reply]
目前ppocr系列开源出来的模型没有针对拼音进行训练，需要你构建一批数据集，微调当前模型，包括对识别字典的更新

**Comment by robotJie** [comment]
@UserWangZz 了解了，数据集的数量有待增加。然后PPChatOCR v3发布了，我试试看。

字典用的ppocr_keys_v1.txt，然后加了几个里头没有的带声调的字母。所以是因为这个导致index变化了，所以模型预测不符合预期？
那是不是说如果要额外增加字典的字符，只能从末尾append去加喽？

**Comment by UserWangZz** [comment]
> @UserWangZz 了解了，数据集的数量有待增加。然后PPChatOCR v3发布了，我试试看。
> 
> 字典用的ppocr_keys_v1.txt，然后加了几个里头没有的带声调的字母。所以是因为这个导致index变化了，所以模型预测不符合预期？ 那是不是说如果要额外增加字典的字符，只能从末尾append去加喽？

是的，因为模型是基于这个字典进行的训练，如果index变化了，就会导致错误

**Comment by robotJie** [comment]
@UserWangZz 我用v3和未修改的ppocr_keys_v1.txt测试上面beizhu.png，能够正常识别了!
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/798af8d6-ff30-47cf-92e7-d6af13699460)

但是只要在ppocr_keys_v1.txt加上自定义的字符，识别结果就非常奇怪。这是正常的么嘛？(已经是append到末尾了)
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/e5a20779-207c-4ee6-9bf8-3ea17d017869)
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/158746475/1b1eac3d-4d81-4214-80d0-2e98dcb6be17)

**Reply by skyler9901** [reply]
请问你在ppocr_key_v1.txt字典里加入自定义字符，如何保持他们说的匹配的呀？还有你后来训练效果怎么样呀，最近也在做微调。

**Reply by UserWangZz** [reply]
匹配其实就是模型输出字典中所有字符的一个概率，因此当字典长度发生改变后，模型输出和字典无法进行一个对比，就会导致输出乱码字符。

**Comment by UserWangZz** [comment]
> @UserWangZz 我用v3和未修改的ppocr_keys_v1.txt测试上面beizhu.png，能够正常识别了! ![image](https://private-user-images.githubusercontent.com/158746475/340650795-798af8d6-ff30-47cf-92e7-d6af13699460.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MTY2MTEsIm5iZiI6MTcxODcxNjMxMSwicGF0aCI6Ii8xNTg3NDY0NzUvMzQwNjUwNzk1LTc5OGFmOGQ2LWZmMzAtNDdjZi05MmU3LWQ2YWYxMzY5OTQ2MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxOFQxMzExNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOTU4YWYxYmMxZjc3YTM4YTdjNGVjYjI3Njc4ZDdhOGUwNDQxZTQ0Y2M0YzBjNzJjZWMwNzg5NDEwYjk0YWNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NUNj4kHJlrzpgaIJOXvYICUfddok6fzqE-9EMRdzP_M)
> 
> 但是只要在ppocr_keys_v1.txt加上自定义的字符，识别结果就非常奇怪。这是正常的么嘛？(已经是append到末尾了) ![image](https://private-user-images.githubusercontent.com/158746475/340651238-e5a20779-207c-4ee6-9bf8-3ea17d017869.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MTY2MTEsIm5iZiI6MTcxODcxNjMxMSwicGF0aCI6Ii8xNTg3NDY0NzUvMzQwNjUxMjM4LWU1YTIwNzc5LTIwN2MtNGVlNi05YmY4LTNlYTE3ZDAxNzg2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxOFQxMzExNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNGUzYWIyNTQxMGJjNDcwZjkzYTYzNDBhMmM3OWIwNGY0M2U3MDc5M2NmN2UxNDMwNDBmMzYzNmE3MGNiODY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.RphmDzJN3SbTkCrHuPHkNohpk1TG1Nc62fXk0h5Hvnw) ![image](https://private-user-images.githubusercontent.com/158746475/340651295-1b1eac3d-4d81-4214-80d0-2e98dcb6be17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MTY2MTEsIm5iZiI6MTcxODcxNjMxMSwicGF0aCI6Ii8xNTg3NDY0NzUvMzQwNjUxMjk1LTFiMWVhYzNkLTRkODEtNDIxNC04MGQwLTJlOThkY2I2YmUxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxOFQxMzExNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Yzc5MjVmNGQ0ZThhNTcyNjc1NzcyNzEwOTM2MDE4OTYzMTJlYmJhY2VmYjY3ZjJkZTlhNGE0YzU1OWI0MWY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.opgbc-yWAl52vRF5zK9euHBtP0Qtlnx74eh7pV7fkOc)

这种情况是字典增多与模型最后的FC层维度不匹配造成的，可以简单微调模型，有能力可以冻住模型参数，只更新FC层尝试
